File: robot_sim/gaussian_field_env.cc
#include "gaussian_field_env.hh"
#include <algorithm>
#include <cmath>
#include <print>

namespace seqcomm {
using namespace std::chrono_literals;

// ── Construction / reset ──────────────────────────────────────────────────────

GaussianFieldEnv::GaussianFieldEnv(Config cfg, random_source& rng)
    : cfg_(cfg), rng_(rng),
      gaussians_(cfg.n_gaussians),
      positions_(cfg.n_agents),
      pending_actions_(cfg.n_agents, -1),
      next_obs_(cfg.n_agents)
{
    reset();
}

std::vector<std::vector<float>> GaussianFieldEnv::reset() {
    const float G = static_cast<float>(cfg_.grid_size);

    // Randomise Gaussian centers and velocities
    for (auto& g : gaussians_) {
        g.x  = rng_.uniform(0.f, G);
        g.y  = rng_.uniform(0.f, G);
        float angle = rng_.uniform(0.f, 6.2831853f);
        g.vx = cfg_.move_speed * std::cos(angle);
        g.vy = cfg_.move_speed * std::sin(angle);
    }

    // Randomise agent positions
    for (auto& p : positions_) {
        p[0] = rng_.uniform(0, cfg_.grid_size - 1);
        p[1] = rng_.uniform(0, cfg_.grid_size - 1);
    }

    // Reset barrier
    submitted_     = 0;
    results_ready_ = false;
    std::fill(pending_actions_.begin(), pending_actions_.end(), -1);

    std::vector<std::vector<float>> obs;
    obs.reserve(cfg_.n_agents);
    for (int i = 0; i < cfg_.n_agents; ++i)
        obs.push_back(obs_for(i));
    return obs;
}

int GaussianFieldEnv::obs_dim() const {
    int w = 2 * cfg_.window_half + 1;
    return 2 + w * w;
}

int GaussianFieldEnv::action_dim() const { return N_ACTIONS; }


// ── Field ─────────────────────────────────────────────────────────────────────

float GaussianFieldEnv::field_at(float row, float col) const {
    float val = 0.f;
    float inv2s2 = 1.f / (2.f * cfg_.gaussian_sigma * cfg_.gaussian_sigma);
    for (auto& g : gaussians_) {
        float dr = row - g.x;
        float dc = col - g.y;
        val += std::exp(-(dr*dr + dc*dc) * inv2s2);
    }
    return val;
}

std::vector<float> GaussianFieldEnv::obs_for(int agent_id) const {
    const float G = static_cast<float>(cfg_.grid_size);
    const int   w = cfg_.window_half;
    int row = positions_[agent_id][0];
    int col = positions_[agent_id][1];

    std::vector<float> o;
    o.reserve(obs_dim());
    o.push_back(row / G);
    o.push_back(col / G);

    for (int dr = -w; dr <= w; ++dr) {
        for (int dc = -w; dc <= w; ++dc) {
            int nr = row + dr, nc = col + dc;
            if (nr >= 0 && nr < cfg_.grid_size && nc >= 0 && nc < cfg_.grid_size)
                o.push_back(field_at(static_cast<float>(nr),
                                     static_cast<float>(nc)));
            else
                o.push_back(0.f);
        }
    }
    return o;
}


// ── Step ──────────────────────────────────────────────────────────────────────

void GaussianFieldEnv::apply_and_step() {
    // Action deltas: stay, up, down, left, right
    constexpr int DR[5] = {0, -1, 1,  0, 0};
    constexpr int DC[5] = {0,  0, 0, -1, 1};
    const int G = cfg_.grid_size;

    for (int i = 0; i < cfg_.n_agents; ++i) {
        int a = pending_actions_[i];
        if (a < 0 || a >= N_ACTIONS) a = 0;
        positions_[i][0] = std::clamp(positions_[i][0] + DR[a], 0, G - 1);
        positions_[i][1] = std::clamp(positions_[i][1] + DC[a], 0, G - 1);
    }

    // Compute reward
    float reward = 0.f;
    for (int i = 0; i < cfg_.n_agents; ++i)
        reward += field_at(static_cast<float>(positions_[i][0]),
                           static_cast<float>(positions_[i][1]));
    for (int i = 0; i < cfg_.n_agents; ++i)
        for (int j = i + 1; j < cfg_.n_agents; ++j)
            if (positions_[i] == positions_[j])
                reward -= cfg_.overlap_penalty;
    step_reward_ = reward;

    // Advance Gaussians (bounce off walls)
    const float Gf = static_cast<float>(G);
    for (auto& g : gaussians_) {
        g.x += g.vx;
        g.y += g.vy;
        if (g.x < 0.f)  { g.x = -g.x;          g.vx = std::abs(g.vx);  }
        if (g.x >= Gf)  { g.x = 2.f*Gf - g.x;  g.vx = -std::abs(g.vx); }
        if (g.y < 0.f)  { g.y = -g.y;           g.vy = std::abs(g.vy);  }
        if (g.y >= Gf)  { g.y = 2.f*Gf - g.y;  g.vy = -std::abs(g.vy); }
    }

    // Build next observations
    for (int i = 0; i < cfg_.n_agents; ++i)
        next_obs_[i] = obs_for(i);
}


// ── Environment interface ─────────────────────────────────────────────────────

void GaussianFieldEnv::submit_action(int agent_id,
                                     std::span<const float> action) {
    // First submission of a new round: clear results from previous round
    if (submitted_ == 0)
        results_ready_ = false;

    pending_actions_[agent_id] = static_cast<int>(action[0]);
    ++submitted_;

    if (submitted_ == cfg_.n_agents) {
        apply_and_step();
        submitted_     = 0;
        results_ready_ = true;
    }
}

cot::task<std::pair<std::vector<float>, float>>
GaussianFieldEnv::get_result(int agent_id) {
    // Yield in a tight loop until all agents have submitted and the step ran.
    // Each co_await gives other cotamer tasks (other agents) a chance to run.
    while (!results_ready_)
        co_await cot::after(1us);
    co_return {next_obs_[agent_id], step_reward_};
}

} // namespace seqcomm

File: execution/gaussian_field_env.py
"""
Python mirror of robot_sim/gaussian_field_env — used during MAPPO training.

Identical dynamics to the C++ version; synchronous since the Python training
loop steps all agents together rather than running them as cotamer tasks.

Observation o_i: [row/G, col/G, field_window_flat]
  - 2 normalised position values
  - (2*window_half+1)^2 field samples centred on agent, zero-padded at edges

Action: integer in {0,1,2,3,4} = stay, up, down, left, right

Reward (shared across all agents at each timestep):
  sum_i F(p_i)  -  overlap_penalty * |{(i,j): p_i==p_j, i<j}|
"""

import numpy as np
from dataclasses import dataclass, field
from typing import List, Tuple


@dataclass
class GaussianFieldConfig:
    grid_size:       int   = 20
    window_half:     int   = 2      # window = (2*window_half+1)^2
    n_gaussians:     int   = 3
    gaussian_sigma:  float = 2.5
    move_speed:      float = 0.4    # Gaussian center displacement per step
    overlap_penalty: float = 1.0    # lambda
    n_agents:        int   = 4


class GaussianFieldEnv:
    """
    2D cooperative coverage environment.

    Agents navigate a grid to collect reward from K moving Gaussian peaks.
    They are penalised for occupying the same cell (duplicate coverage).
    The hidden field is never directly revealed — agents only see a local
    window, so the world model M must learn to predict field evolution.
    """

    N_ACTIONS = 5  # stay, up, down, left, right
    _DELTAS = [(0, 0), (-1, 0), (1, 0), (0, -1), (0, 1)]

    def __init__(self, cfg: GaussianFieldConfig = GaussianFieldConfig(),
                 seed: int | None = None):
        self.cfg = cfg
        self.rng = np.random.default_rng(seed)
        self._win = 2 * cfg.window_half + 1
        # state initialised in reset()
        self.centers    = np.zeros((cfg.n_gaussians, 2), dtype=np.float32)
        self.velocities = np.zeros((cfg.n_gaussians, 2), dtype=np.float32)
        self.positions  = np.zeros((cfg.n_agents, 2), dtype=np.int32)
        self.reset()

    # ── Dimensions ────────────────────────────────────────────────────────────

    @property
    def obs_dim(self) -> int:
        return 2 + self._win ** 2

    @property
    def action_dim(self) -> int:
        return self.N_ACTIONS

    # ── Lifecycle ─────────────────────────────────────────────────────────────

    def reset(self) -> List[np.ndarray]:
        """Reset field and agent positions. Returns list of obs per agent."""
        G = self.cfg.grid_size
        self.centers = self.rng.uniform(0, G, (self.cfg.n_gaussians, 2)).astype(np.float32)
        angles = self.rng.uniform(0, 2 * np.pi, self.cfg.n_gaussians)
        self.velocities = (self.cfg.move_speed *
                           np.stack([np.cos(angles), np.sin(angles)], axis=1)
                           .astype(np.float32))
        self.positions = self.rng.integers(0, G, (self.cfg.n_agents, 2)).astype(np.int32)
        return [self._obs_for(i) for i in range(self.cfg.n_agents)]

    def step(self, actions: List[int]) -> Tuple[List[np.ndarray], float, bool]:
        """
        Step the environment.

        Args:
            actions: list of int (one per agent), each in [0, N_ACTIONS)
        Returns:
            (obs_list, reward, done)
            done is always False — episodes end by timestep count
        """
        G = self.cfg.grid_size

        # Move agents
        for i, a in enumerate(actions):
            dr, dc = self._DELTAS[int(a)]
            self.positions[i, 0] = np.clip(self.positions[i, 0] + dr, 0, G - 1)
            self.positions[i, 1] = np.clip(self.positions[i, 1] + dc, 0, G - 1)

        # Reward: field value at each agent's position
        reward = sum(
            self._field_at(self.positions[i, 0], self.positions[i, 1])
            for i in range(self.cfg.n_agents)
        )

        # Overlap penalty
        for i in range(self.cfg.n_agents):
            for j in range(i + 1, self.cfg.n_agents):
                if np.array_equal(self.positions[i], self.positions[j]):
                    reward -= self.cfg.overlap_penalty

        # Advance Gaussians and bounce off walls
        self.centers += self.velocities
        for k in range(self.cfg.n_gaussians):
            for d in range(2):
                if self.centers[k, d] < 0:
                    self.centers[k, d] *= -1
                    self.velocities[k, d] = abs(self.velocities[k, d])
                elif self.centers[k, d] >= G:
                    self.centers[k, d] = 2 * G - self.centers[k, d]
                    self.velocities[k, d] = -abs(self.velocities[k, d])

        obs = [self._obs_for(i) for i in range(self.cfg.n_agents)]
        return obs, float(reward), False

    # ── Internals ─────────────────────────────────────────────────────────────

    def _field_at(self, row: float, col: float) -> float:
        inv2s2 = 1.0 / (2.0 * self.cfg.gaussian_sigma ** 2)
        dr = row - self.centers[:, 0]
        dc = col - self.centers[:, 1]
        return float(np.exp(-(dr * dr + dc * dc) * inv2s2).sum())

    def _obs_for(self, agent_id: int) -> np.ndarray:
        G   = self.cfg.grid_size
        w   = self.cfg.window_half
        row = int(self.positions[agent_id, 0])
        col = int(self.positions[agent_id, 1])

        obs = [row / G, col / G]
        for dr in range(-w, w + 1):
            for dc in range(-w, w + 1):
                nr, nc = row + dr, col + dc
                if 0 <= nr < G and 0 <= nc < G:
                    obs.append(self._field_at(nr, nc))
                else:
                    obs.append(0.0)
        return np.array(obs, dtype=np.float32)

    # ── Rendering (optional) ──────────────────────────────────────────────────

    def render_field(self) -> np.ndarray:
        """Return the full field as a (grid_size, grid_size) array for viz."""
        G = self.cfg.grid_size
        grid = np.zeros((G, G), dtype=np.float32)
        for r in range(G):
            for c in range(G):
                grid[r, c] = self._field_at(r, c)
        return grid
